### PR TITLE
feat: add Gemma 4 unified audio input support

### LINKED ISSUE #591: Feature Request: Audio input support in /v1/chat/completions for multimodal models (e.g. Gemma-4)
# Feature Request: Audio input support in /v1/chat/completions for multimodal models (e.g. Gemma-4)

**Is your feature request related to a problem? Please describe.**

When using multimodal models like `gemma-4-e2b-it-4bit` or `gemma-4-e4b-it-4bit` that natively support audio input, it is currently not possible to pass audio data through the `/v1/chat/completions` endpoint. Sending audio via the `input_audio` content type (OpenAI-compatible format) results in the model receiving only the text portion of the message — the audio is silently ignored and never processed.

For example, the following request results in the model responding as if no audio was provided (only 17 prompt tokens, no audio tokens):

```bash
curl -X POST http://localhost:8005/v1/chat/completions \
  -H "Authorization: Bearer <key>" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "gemma-4-e2b-it-4bit",
    "messages": [
      {
        "role": "user",
        "content": [
          {"type": "text", "text": "Please describe this audio"},
          {
            "type": "input_audio",
            "input_audio": {
              "data": "<base64-encoded-wav>",
              "format": "wav"
            }
          }
        ]
      }
    ]
  }'
```

The model replies asking the user to provide audio, meaning the audio content block was not forwarded to the model at all.

**Describe the solution you'd like**

Support the `input_audio` content type in `/v1/chat/completions` for models that have audio input capability (e.g. Gemma-4 series). The audio should be decoded from base64 and passed to the model's processor alongside the text tokens, similar to how image input is already handled for vision models.

Ideally, this would follow the OpenAI audio input format:

```json
{
  "type": "input_audio",
  "input_audio": {
    "data": "<base64-encoded-audio>",
    "format": "wav"
  }
}
```

Supported formats should include at minimum `wav` and `mp3`.

**Describe alternatives you've considered**

1. **Using `/v1/audio/transcriptions` first, then passing text to chat completions** — This works as a workaround (e.g. using `Qwen3-ASR-0.6B` for transcription), but it requires two separate API calls, increases latency, and loses any non-verbal audio information that the multimodal model could otherwise interpret directly.

2. **Using `mlx_vlm.generate --audio` CLI directly** — This works perfectly (verified with `gemma-4-e4b-it-4bit`), confirming the underlying model supports audio input. The gap is only at the HTTP server layer.

**Additional context**

- Verified that `mlx_vlm.generate --audio test.wav --model gemma-4-e4b-it-4bit` correctly processes audio, proving the model itself supports audio input natively.
- The `/v1/audio/transcriptions` endpoint works correctly for dedicated ASR models (e.g. `Qwen3-ASR-0.6B`).
- This feature would unlock the full multimodal capability of Gemma-4 and similar models through the HTTP API, making it consistent with the CLI behavior already supported by `mlx_vlm`.

### PR DESCRIPTION
**Pull request by sje397**
- Support audio input alongside images in VLM engine (BytesIO → numpy, has_multimodal gate now includes audio, image code guarded with num_images > 0)
- Add gemma4_unified model type throughout adapter, API, and server layers
- Add suppress_tokens mechanism in scheduler to prevent emission of multimodal placeholder tokens (<image|>, <audio|>)
- Add end-of-turn token detection for Gemma 4 (<turn|>, ID 106) to prevent hallucinating past conversational turns
- Add PIL feature extractor fallback for audio processor compatibility
- Extract audio data from input_audio content blocks in messages
- Bump mlx-vlm pin to 526c210 (includes Gemma4 unified prefill fixes)
- Add tests for audio input, suppress tokens, scheduler, and VLM engine

Add Gemma 4 unified audio input support to the Anthropic and OpenAI endpoints, enabling audio alongside images and text in multimodal requests. Extends the VLM engine to process audio data alongside images, adds suppress-token logic for multimodal placeholder tokens, and adds end-of-turn detection for Gemma 4.

## Dependency Chain

**Upstream mlx-vlm pin: `041f889` → `526c210`**
Picks up [mlx-vlm#1292](https://github.com/Blaizzy/mlx-vlm/pull/1292) — video input support for Gemma 4 12B, which includes the Gemma4 unified prefill fixes needed for audio. Also includes #1291 (APC fix for single requests) and unrelated commits for PaddleOCR/Nemotron/Ideogram.

### Anthropic Endpoint (`/v1/messages`)
- **`ContentBlockInputAudio`** model: `data` (base64) + `format` (default `"wav"`)
- **`convert_anthropic_to_internal()`** extended to preserve `input_audio` blocks when `preserve_images=True`, extracting audio data for VLM processing
- **`_build_message_from_parts()`** handles `input_audio` alongside images

### OpenAI Endpoint (`/v1/chat/completions`)
- **`InputAudio`** model: `data` (base64 or data URI) + `format` (default `"wav"`)
- **`ContentPart.type`** extended to accept `"input_audio"` with `InputAudio` field
- Audio content parts parsed through the same internal pipeline as Anthropic

### VLM Engine
- `has_multimodal` gate now includes audio (not just images)
- Image processing guarded with `num_images > 0` to avoid empty-image errors with audio-only requests
- Audio data extracted from `input_audio` content blocks in messages → BytesIO → numpy
- PIL feature extractor fallback for audio processor compatibility

### Scheduler
- **Suppress-tokens mechanism** prevents emission of multimodal placeholder tokens (`<image|>`, `<audio|>`) that would otherwise leak through logits post-processing
- **End-of-turn detection** for Gemma 4 (`<turn|>`, ID 106) prevents hallucinating past conversational turns

### Model Adapter
- `gemma4_unified` model type added throughout adapter, API, and server layers

## Test Coverage

| Layer | File | Tests | Coverage |
|-------|------|-------|----------|
| Anthropic conversion | `test_anthropic_adapter.py` | 3 | `ContentBlockInputAudio` preserve/drop/audio-only |
| API utils (shared) | `test_api_utils.py` | 4 | Audio content block extraction from messages |
| Image/audio utils | `test_image_utils.py` | 6 | Base64, raw bytes, paths, data URIs, mixed |
| **OpenAI models** | `test_openai_models.py` | **9** | `InputAudio` creation/validation/serialization; `ContentPart` with `type="input_audio"`; `Message` with mixed audio+text |
| **OpenAI adapter** | `test_openai_adapter.py` | **2** | `parse_request` with audio content (mixed + audio-only), verifies no crash |
| VLM engine | `test_vlm_engine.py` | ~177 lines | Audio pipeline, multimodal gate, BytesIO→numpy |
| Scheduler | `test_scheduler.py` | ~36 lines | Suppress tokens, end-of-turn detection |
| Gemma4 messages | `test_gemma4_messages.py` | ~16 lines | `<turn|>` token handling |

## File Stats

```
18 files changed, +1054, -66
```

**Production code (8 files):** `anthropic_models.py` (+13), `anthropic_utils.py` (+59/-?), `openai_models.py` (+12), `api/utils.py` (+24/-?), `engine/vlm.py` (+152/-?), `scheduler.py` (+90), `utils/image.py` (+57/-?), `adapter/gemma4.py` (+9/-?)

**Tests (8 files):** `test_anthropic_adapter.py` (+123), `test_api_utils.py` (+43), `test_image_utils.py` (+140/-?), `test_vlm_engine.py` (+177/-?), `test_scheduler.py` (+36), `test_gemma4_messages.py` (+16), `test_openai_models.py` (+113), `test_openai_adapter.py` (+45)

**Config (1 file):** `pyproject.toml` (mlx-vlm pin bump)

Fixes https://github.com/jundot/omlx/issues/591

### CODE REVIEW

**Comment by jundot** [comment]
Thanks for the PR. I reviewed the audio input path for OpenAI/Anthropic, the VLM preprocessing changes, suppress-token handling, and Gemma 4 turn stop behavior; the focused tests and a local Gemma 4 12B audio smoke test passed for me.

The CI failure is just a small test fixture gap in a partial Scheduler mock, not a runtime issue with this change. This looks good to me, and I'm going to merge it; I'll fold that test fix into a follow-up.

**Comment by sje397** [comment]
You're very welcome. Thanks for merging, and thanks for oMLX.

**Comment by monroewilliams** [comment]
Thanks for getting this in! I was working on similar changes for a local experiment, but I wasn't handling the end-of-turn token proerly and was getting some misbehavior from the model. Your changes cleaned all that up. :)

I did run into one issue with dependency imports, which I've filed as https://github.com/jundot/omlx/issues/1688

**Comment by ellekdev** [comment]
Thanks for this PR! Super cool! I was wondering if it'd be possible to also add similar support in the OMLX chat UI for audio files? Currently, when I try to upload audio in the chat UI, I get `Only PDF, DOCX, PPTX, TXT, and MD documents are supported.`

Maybe there is a setting I'm misconfiguring? Or is this not currently supported?

Thanks!
